ODIN_II/SRC/netlist_create_from_ast.cpp

```diff
@@ -3469,51 +3472,213 @@ signal_list_t* create_operation_node(ast_node_t* op, signal_list_t** input_lists
+                }
                 /* allocate the pins needed */
                 allocate_more_input_pins(operation_node, input_port_width);
                 /* record this port size */
                 add_input_port_information(operation_node, input_port_width);
                 /* hookup the input pins */
-                hookup_input_pins_from_signal_list(operation_node, current_idx, input_lists[i], 0, input_port_width, verilog_netlist);
+                /* 
+                *   if the input is signed, then instead of adding pad 0 for extension we have to add its sign bit as extension bits
+                */
+                hookup_input_pins_from_signal_list(operation_node, current_idx, shift_input_pins[i], 0, input_port_width, verilog_netlist);
+                
+                if (op->children[1]->type != NUMBERS)
+                    current_idx += input_port_width;
             }
-        } else if (input_port_width != -2) {
-            /* IF taking port widths based on preset */
-            /* allocate the pins needed */
-            allocate_more_input_pins(operation_node, input_port_width);
-            /* record this port size */
-            add_input_port_information(operation_node, input_port_width);
+            if ((i == 1) && (op->children[1]->type != NUMBERS)) {
+                //
+                for (int j = 0; j < input_lists[i]->count; j++) {
+                    add_pin_to_signal_list(shift_input_pins[i], input_lists[i]->pins[j]);
+                    input_lists[i]->pins[j] = NULL;
+                }
+                /* allocate the pins needed */
+                allocate_more_input_pins(operation_node, log2(op->children[1]->types.vnumber->get_value()));
+                /* record this port size */
+                add_input_port_information(operation_node, log2(op->children[1]->types.vnumber->get_value()));
+                /* hookup the input pins */
+                hookup_input_pins_from_signal_list(operation_node, current_idx, shift_input_pins[i], 0, log2(op->children[1]->types.vnumber->get_value()), verilog_netlist);
+            }
+        }
 
-            /* hookup the input pins = will do padding of zeros for smaller port */
-            hookup_input_pins_from_signal_list(operation_node, current_idx, input_lists[i], 0, input_port_width, verilog_netlist);
-            current_idx += input_port_width;
-        } else {
-            /* ELSE if taking the port widths as they are */
-            /* allocate the pins needed */
-            allocate_more_input_pins(operation_node, input_lists[i]->count);
-            /* record this port size */
-            add_input_port_information(operation_node, input_lists[i]->count);
+        if (op->children[1]->type == NUMBERS) {
 
-            /* hookup the input pins */
-            hookup_input_pins_from_signal_list(operation_node, current_idx, input_lists[i], 0, input_lists[i]->count, verilog_netlist);
+            /* allocate the pins for the ouput port */
```

Comment:
this should not be done twice (see bellow the case switch, you are doing it again)